velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -563,6 +565,81 @@ TpchPlan TpchQueryBuilder::getQ10Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ12Plan() const {
+  std::vector<std::string> ordersColumns = {"o_orderkey", "o_orderpriority"};
+  std::vector<std::string> lineitemColumns = {
+      "l_receiptdate",
+      "l_orderkey",
+      "l_commitdate",
+      "l_shipmode",
+      "l_shipdate"};
+
+  auto ordersSelectedRowType = getRowType(kOrders, ordersColumns);
+  const auto& ordersFileColumns = getFileColumnNames(kOrders);
+  auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId ordersScanNodeId;
+  core::PlanNodeId lineitemScanNodeId;
+
+  const std::string receiptDateFilter = formatDateFilter(
+      "l_receiptdate", lineitemSelectedRowType, "'1994-01-01'", "'1994-12-31'");
+  const std::string shipDateFilter = formatDateFilter(
+      "l_shipdate", lineitemSelectedRowType, "", "'1995-01-01'");
+  const std::string commitDateFilter = formatDateFilter(
+      "l_commitdate", lineitemSelectedRowType, "", "'1995-01-01'");
+  const std::string commitReceiptDateFilter = "l_commitdate < l_receiptdate";
+  const std::string shipCommitDateFilter = "l_shipdate < l_commitdate";
+  const std::string shipModeFilter = "l_shipmode IN ('MAIL', 'SHIP')";
+
+  auto lineitem = PlanBuilder(planNodeIdGenerator, pool_.get())
+                      .tableScan(
+                          kLineitem,
+                          lineitemSelectedRowType,
+                          lineitemFileColumns,
+                          {receiptDateFilter,
+                           shipModeFilter,
+                           shipDateFilter,
+                           commitDateFilter},
+                          commitReceiptDateFilter)
+                      .capturePlanNodeId(lineitemScanNodeId)
+                      .filter(shipCommitDateFilter)
+                      .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kOrders, ordersSelectedRowType, ordersFileColumns, {})
+          .capturePlanNodeId(ordersScanNodeId)
+          .project(
```

Comment:
Isn't it better to compute this after the join?

Reply:
In general, the selectivity of the join is a runtime property. But looking at Q12, I would expect the optimizer to compute this after the join.

Reply:
Thanks for the suggestion, I looked at the detailed plan and Presto is doing this as well, the times have improved slightly with this optimization and I have updated the PR with new times